Improved SoC handling provided by chargers

DerAndereAndi opened the pull request:
- If the charger can not provide SoC with the current car connection, it has to return `api.ErrNotAvailable`
- If the charger does not know yet if it can provide SoC with the current car connection, it has to return `api.ErrMustRetry`
- If the charger does or could provide an SoC, the vehicle API will never be called, even in case of an error

DerAndereAndi commented:
Das sollte die Implementierung von https://github.com/evcc-io/evcc/pull/1408#issuecomment-903280791 sein

andig commented:
Ich glaub ein Testcase wäre gut der auch den Fallback aufs Fahrzeug zeigt

DerAndereAndi commented:
Yep. Hab ich noch vor. Morgen dann :)

DerAndereAndi committed:
Added test case

DerAndereAndi commented:
Morgen ist doch schon heute...

andig commented on charger/eebus.go:
```diff
@@ -505,7 +505,7 @@ func (c *EEBus) SoC() (float64, error) {
 	data, err := c.cc.GetData()
 	if err != nil {
 		c.log.TRACE.Printf("soc: no eebus data available yet")
```
Fehlerbehandlung nur 1x: entweder loggen ODER zurück geben

DerAndereAndi replied:
Ich habe das momentan mehrfach so gemacht, damit ich den Ablauf in den Trace Logs besser nachvollziehen kann und so schneller etwaige Probleme finde. Daher würde ich das so lange es `Experimental` ist gerne drin lassen.

andig replied:
Ich hab die experimentellen Logs mit !! gestartet, die finde ich dan später schnell beim rausschmeissen. Wäre vllt auch für Dich eine Option

andig commented on core/soc/estimator.go:
```diff
@@ -98,7 +98,23 @@ func (s *Estimator) SoC(chargedEnergy float64) (float64, error) {
 
-		if err == nil {
+		// if the charger does or could provide SoC, we always use it instead of using the vehicle API
+		if err == nil || !errors.Is(err, api.ErrNotAvailable) {
```
Warum Test auf ErrNotAvailable- wo kann der her kommen?

andig commented on core/soc/estimator.go:
```diff
@@ -98,7 +98,23 @@ func (s *Estimator) SoC(chargedEnergy float64) (float64, error) {
+		// if the charger does or could provide SoC, we always use it instead of using the vehicle API
+		if err == nil || !errors.Is(err, api.ErrNotAvailable) {
+			if err != nil {
+				if errors.Is(err, api.ErrMustRetry) {
```
Warum Abbruch bei ErrMustRetry? Im nächsten Absatz nehmen wir dann einfach den letzten SoC- warum nicht auch hier?

DerAndereAndi replied:
Du meinst i `if s.prevSoC == 0 {...` wird das schon abgefangen? Jo, ist dann beim Code drunter für Vehicle API identisch. Kann also bei beiden raus

DerAndereAndi replied:
Habs mal so gemacht.

DerAndereAndi committed:
Mark experimental trace logs

DerAndereAndi committed:
Update tests with errors returning soc 0

DerAndereAndi committed:
Remove not needed error checks

andig commented:
/rebase

DerAndereAndi commented:
rebase done

andig commented:
/rebase

DerAndereAndi committed:
Improved SoC handling provided by chargers

- If the charger can not provide SoC with the current car connection, it has to return `api.ErrNotAvailable`
- If the charger does not know yet if it can provide SoC with the current car connection, it has to return `api.ErrMustRetry`
- If the charger does or could provide an SoC, the vehicle API will never be called, even in case of an error

DerAndereAndi committed:
Fix test

andig commented on core/soc/estimator.go:
```diff
@@ -107,10 +119,6 @@ func (s *Estimator) SoC(chargedEnergy float64) (float64, error) {
 	if fetchedSoC == nil {
 		f, err := s.vehicle.SoC()
 		if err != nil {
-			if errors.Is(err, api.ErrMustRetry) {
```
damit wir errmustretry jetzt immer verschlucket- für apis mit refreshkey ein nogo.

DerAndereAndi replied:
Also bei Chargern eine andere Logik als bei Vehicle API? Warum? Wie macht man das im Code klar warum?

DerAndereAndi replied:
Habs beim Charger mal entfernt, in der Vermutung dass wir das nicht brauchen werden. (Keine Ahnung ob das nicht mal anders sein kann)

Und für die Vehicle API einen Kommentar hinzugefügt, dass man auch weiss warum das dort so ist. Denn von der Logik ist das nicht klar.

DerAndereAndi committed:
Revert "Remove not needed error checks"

This reverts commit 732d78d7d96f9ed5c58790c4449e3eb73d3c62d2.

DerAndereAndi committed:
Update ErrMustRetry handling

- Charger likely don't need that, so remove it
- Vehicle Online APIs need that because of the use of refresh tokens

andig commented on core/soc/estimator.go:
```diff
@@ -107,6 +119,7 @@ func (s *Estimator) SoC(chargedEnergy float64) (float64, error) {
 	if fetchedSoC == nil {
 		f, err := s.vehicle.SoC()
 		if err != nil {
+			// required for online APIs with refresh tokens
```
ein refresh token ist egtl etwas anderes

DerAndereAndi replied:
Also `// required for online APIs with refreshkey` ?

DerAndereAndi committed:
Update comment
